### ISSUE TITLE
Private branches for public repositories

### ISSUE THREAD

**Issue by erkinalp** [isaacs/github#499]
Some open source projects have companion proprietary projects sharing lots of code and issues. Along with appropriate access control in issues and PRs, this will help deduplication.

**Comment by cirosantilli** [comment]
Does the proprietary code have issues from users which cannot see the source?

Would you also need a per issue public / private permission?

**Comment by erkinalp** [comment]
Yes. However, developers will have to care not to reveal closed source code in a public comment.

**Comment by Lemongrass3110** [comment]
+1 for this. At times like these it would be better to have a feature like this(PR/branches for non public exploits).

**Comment by nottrobin** [comment]
I really want this feature. But neither Bitbucket not Gitlab seem to offer it, so I'm wondering if it's deeply incompatible with the way git works?

**Comment by ProLoser** [comment]
I think it should be private forks of public repos, but that may have some conflicts with licensing issues.

**Comment by ChrisRackauckas** [comment]
This is really big for academics who cannot release code until after publication, but may be implemented a new algorithm by modifying an existing one (especially when it's a new addition to one's own specialized package).

**Comment by ProLoser** [comment]
While you can't really fork and change privacy, you CAN 'fake' a fork by simply creating a 'new' private repo and pushing all of the public history to it. You lose the cross-repo Pull-Request GUI but you can still use GIT locally to do all the updating. It's the same as having 2 remotes (one is public, one is private). Just make sure not to push to the public one.

**Comment by brittanystoroz** [comment]
I would absolutely die and go to heaven if private branches became a thing. I'm writing curriculum for Turing School and a lot of our lessons come with tutorial repos where we want a public 'begin' branch that just contains boilerplate for students to build off of to complete the tutorial. It would be amazing to have a private branch for the completed version so that instructors could use it for reference, but students wouldn't be able to look ahead and copy the completed example instead of doing the work.

**Comment by Alex-Tideman** [comment]
I also teach at Turing School of Software and WOULD LOVE THIS FEATURE. Having the ability to store private branches on a repo for answers would be so nice so our projects stay well organized, we won't need a million repos, and the students can't get free answers.

**Comment by jsocol** [comment]
I used to maintain an open source security tool, private issues and branches would have been _huge_, since currently any issue, branch, or PR is potentially immediate disclosure of a serious issue.

**Comment by stacimcwilliams** [comment]
👍

**Comment by tmikeschu** [comment]
YAS

**Comment by dinvlad** [comment]
One other use case is maintaining WIP branches, where different developers stash their code when they leave for the day (perhaps to continue working from home ;) ). Currently, they'd have to maintain a separate private repo for that (paid if hosted on GitHub).

**Comment by snmihuss** [comment]
such a useful feature... it's really sad that it is not offered / supported :(

**Comment by micahstubbs** [comment]
👎 feel like this confuses the community about what is open and what is not.  

separate private repos and build tools to combine them with OSS dependencies is the solution I've seen used in practice.

**Comment by justindocanto** [comment]
Absolutely need this. 

For example, right now i need to add a branch for our dev-only tools & reporting for a project that aren't intended to be public. Yes we can make another repo but the tools integrate heavily with the master branch and it would be ideal to have it simply be a branch

**Comment by afbarnard** [comment]
I am also thinking about something like this from a research point of view where code would be kept private until publication. I would be fine with a public fork downstream from a private repo if that would work. This would also seem to be a solution for the freemium model of software. Or maybe it should be the other way around: a private fork downstream from a public repo. Does anybody have experience with this?

**Comment by paaddyy** [comment]
I agree, it would be awesome if you could create private branches on public repos 😍 
`git checkout -b -p XXX` (e.g. a flag for private) would be an amazing command imo 😄

**Comment by eosfor** [comment]
+1

**Comment by grant** [comment]
This would help us with sharing early-access code for new @gsuitedevs APIs with developers before the API is launched.

All of our code is organized by language (i.e. node-samples). We currently have private ({product}-api-samples) repos, but we'll have split these into our 11 public language-based sample repos when the API is public.

Although, I could see how the ACL controls could be a bit confusing.

**Comment by marcussacana** [comment]
Well, Nothing yet?

**Comment by nottrobin** [comment]
@paaddyy When you do `git checkout -b` you're creating a branch within the local git repository in your folder on your computer. At that point, it's not shared anywhere. The cloned repository on your computer has no concept of public or private. Fundamentally, Git, the command-line tool, has no concept of GitHub, and therefore also has no understanding of public or private.

Only once you push your repository to GitHub does public/private have any meaning.

The place where the `git` CLI command and GitHub functionality intersect is [the `hub` project](https://github.com/github/hub). If GitHub were to ever support private branches, it would then probably make sense for `hub` to add functionality to interact with private branches.

**Comment by thorade** [comment]
https://twitter.com/github/status/1131473309470789632

**Comment by itsvinayak** [comment]
really in need of this

**Comment by Erik-Roth** [comment]
This would be a great feature!!! +1

**Comment by saart** [comment]
Yeah!
+1 to that!

**Comment by adithyabsk** [comment]
+1

**Comment by antoniocavalcante** [comment]
+1

**Comment by JamesNewton** [comment]
We really need this!

**Comment by Zero2Cool** [comment]
This would be pretty nice to have.

**Comment by nicolay-zlobin** [comment]
+1

**Comment by OzzieFZI** [comment]
+1

**Comment by ma-sadeghi** [comment]
+1

**Comment by jamesonwilliams** [comment]
I also think that this would be groovy.

**Comment by singlerider** [comment]
👍

**Comment by a17fellu** [comment]
This would be a great addition

+1

**Comment by koulanurag** [comment]
+1

**Comment by xyele** [comment]
+1

**Comment by bohdantverdyi** [comment]
+1

**Comment by mikkelee** [comment]
+1

**Comment by ThizThizzyDizzy** [comment]
+1

**Comment by abhishekyana** [comment]
+1
Desperately need this feature.

**Comment by ObscureBug** [comment]
+1 Really need this feature.

Tools are supposed to make life easier!  Not having this feature makes development harder, particularly on dependent packages, as it requires extra private repos, overhead and tracking - which is all just a waste of valuable time.

**Comment by RashminDungrani** [comment]
+1

**Comment by alexcombessie** [comment]
also interested in this feature

**Comment by catmade** [comment]
+1

**Comment by ObscureBug** [comment]
Sadly, this enhancement didn't make it to the publicly announced product roadmap.

**Comment by MaxKlaxxMiner** [comment]
I have been paying for github pro for several years now. Originally for the feature "private repositories" and now simply as a donation. There are several "branch protection rules" in the settings here, but "private branches" would still fit nicely. :-)

For example, it is a bit tedious to copy test photos manually between home, office and members because there is no final copyright release yet. (it is not yet clear which pictures will be published in the future)

+1

**Comment by DarthXoc** [comment]
+1, can't believe this feature doesn't exist already!

**Comment by ma-sadeghi** [comment]
+1

**Comment by Jighdan** [comment]
+1

**Comment by balupton** [comment]
Another rationale for this that hasn't been mentioned above, is keeping public discovery of repositories (name, description, topics, stars, watching, issues), yet keeping access to source code private. Such use cases for this includes incentive mechanisms to protect the source code from exploitation. Such as access to source code only if you sign a contract/code-of-conduct/license/nda/cla/or-whatever, or access to source code only if you are a sponsor. This could empower fairness again from the open-and-be-exploited dynamic of current FLOSS, to open-to-those-who-won't-exploit-us. Such exploitation is plagiarising of code or benefit without return, which exploits the labour of the original producer, often for the most popular plagiariser.

More details on this model here: https://forum.artlessdevices.com/t/coalition-for-access/90?u=balupton

**Comment by selimrecep** [comment]
+1

**Comment by ehorodyski** [comment]
I'd love this feature. I plan on writing some projects that would have a base free OSS main branch, where people can fork the public portion of the code, but additional features would be sold -- therefore private. 

I'm not going to make a living on what I have planned, but would like something more concrete than donations for people who want to support my projects to get

**Comment by erkinalp** [comment]
In fact, the most prominent user would be Microsoft itself, with Microsoft/Windows repository.

**Comment by alimpk** [comment]
This feature is very practical and essential. I also have some academic and industrial projects that part of of them are public and part of them are private. for example, an academic project that provides material for programming assignments. any student needs to fork the project, but requires a private branch to advance solutions. at the moment we have to fork the project twice (public and private).

**Comment by nickyrong** [comment]
+1

**Comment by sbbzplt** [comment]
I need this desperately!

**Comment by fire1ce** [comment]
+1

**Comment by adibrastegarnia** [comment]
+1

**Comment by criptych** [comment]
> In fact, the most prominent user would be Microsoft itself, with https://github.com/Microsoft/Windows repository.

Dunno, it might get inundated by issues... great way for (more tech-savvy) users to report them, though!

**Comment by Prajwalprakash3722** [comment]
+1

**Comment by raiomarco** [comment]
+1
